File: LeetCode/Study Plan/Practice Problems/Easy/Verifying An Alien Dictionary/Java/verifyingAnAlienDictionary.java
class Solution {
    HashMap<Character, Integer> orderNum = new HashMap<>();

    public boolean isAlienSorted(String[] words, String order) {
        int k = 0;
        for (Character c : order.toCharArray()) {
            orderNum.put(c, k++);
        }

        int size = words.length;
        for (int i=1; i < size; i++) {
            String w1 = words[i-1];
            String w2 = words[i];
            if (!isSorted(w1, w2))
                return false;
        }
        return true;
    }

    public boolean isSorted(String s1, String s2) {
        int i=0, j=0;
        
        while (i<s1.length() && j<s2.length() && s1.charAt(i) == s2.charAt(j)) {
            i++;
            j++;
        }

        if (((i<s1.length()) && (j == s2.length())) || 
        ((i<s1.length() && j<s2.length()) && (orderNum.get(s1.charAt(i)) > orderNum.get(s2.charAt(j)))))
            return false;

        return true;
    }
}

File: LeetCode/Study Plan/Practice Problems/Easy/Verifying An Alien Dictionary/Go/verifyingAnAlienDictionary.go
package main

import "fmt"

func isAlienSorted(words []string, order string) bool {
	k := 0
	orderNum := make(map[byte]int)
	for i := 0; i < len(order); i++ {
		orderNum[order[i]] = k
		k++
	}

	size := len(words)
	for k := 1; k < size; k++ {
		w1, w2 := words[k-1], words[k]
		i, j := 0, 0
		for i < len(w1) && j < len(w2) && w1[i] == w2[j] {
			i++
			j++
		}
		if (i < len(w1) && j == len(w2)) || (i < len(w1) && j < len(w2) && orderNum[w1[i]] > orderNum[w2[j]]) {
			return false
		}
	}
	return true
}

func main() {
	fmt.Println(isAlienSorted([]string{"hello", "leetcode"}, "hlabcdefgijkmnopqrstuvwxyz"))
}
